### PR TITLE
Add Dorgtr and test

### PR DESCRIPTION
**Pull request by btracey**

### CODE REVIEW

**Comment by btracey** [comment]
I'd be happy to modify the Dsytrd tests if anyone understands why the Dorgtr test hits the bug but the Dsytrd tests do not. It's not even really specifically a bug with the Dsytrd test... copying and pasting the failing case into Dsytrd made the test fail on that specific case.

File: native/dorgtr.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import "github.com/gonum/blas"
+
+// Dorgtr generates a real orthogonal matrix Q which is defines as the product
```
Comment (vladimir-ch) [inline native/dorgtr.go]: s/defines/defined/

File: native/dorgtr.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+// of n-1 elementary reflectors of order N as returned by Dsytrd.
+//
+// If uplo == blas.Upper
+//  Q = H[n-1] * ... * H[1] * H[0]
+// If uplo == blas.Lower
+//  L = H[0] * ... * H[n-1]
+//
```
Comment (vladimir-ch) [inline native/dorgtr.go]: Could you delete this line and connect the ifs with where to read as a single sentence?

File: native/dorgtr.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+// Where H[i] is constructed from the elementary reflectors as computed by Dsytrd.
+// Please see the documentation for Dsytrd for more information.
```
Comment (vladimir-ch) [inline native/dorgtr.go]: Without `Please`?

File: native/dorgtr.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import "github.com/gonum/blas"
+
+// Dorgtr generates a real orthogonal matrix Q which is defines as the product
+// of n-1 elementary reflectors of order N as returned by Dsytrd.
```
Comment (vladimir-ch) [inline native/dorgtr.go]: Is `n` and `N` the same number?

File: native/dorgtr.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	if len(tau) < n-1 {
+		panic(badTau)
+	}
+	upper := uplo == blas.Upper
+	var nb int
+	if upper {
+		nb = impl.Ilaenv(1, "DORGQL", " ", n-1, n-1, n-1, -1)
+	} else {
+		nb = impl.Ilaenv(1, "DORGQR", " ", n-1, n-1, n-1, -1)
+		nb = 4
```
Comment (vladimir-ch) [inline native/dorgtr.go]: Why do you overwrite `nb` with 4?

File: native/dorgtr.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+	}
+	lworkopt := max(1, n-1) * nb
+	work[0] = float64(lworkopt)
+	if lwork == -1 {
+		return
+	}
+	if len(work) < lwork {
```
Comment (vladimir-ch) [inline native/dorgtr.go]: Swap this if with the next one? Let's first check that `lwork` has a good value and then check that work has a good length.

**Comment by kortschak** [comment]
>I'd be happy to modify the Dsytrd tests if anyone understands why the Dorgtr test hits the bug but the Dsytrd tests do not. It's not even really specifically a bug with the Dsytrd test... copying and pasting the failing case into Dsytrd made the test fail on that specific case.

What do dsytrd.f and dorgtr.f do with the failing case?

This dependence on PRNG sequence is troubling - we should make all the tests truly independent (this means each test uses its own rand.Rand since tests may be run in parallel - so merely doing rand.Seed(1) at the beginning is not adequate). Probably having a test flag that specifies the number of test interations would be a worthwhile addition too.

(I thought I had made this comment on this PR over the weekend, but obviously not).

File: native/dorgtr.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+//  Q = H[n-1] * ... * H[1] * H[0]
+// If uplo == blas.Lower
+//  L = H[0] * ... * H[n-1]
+//
+// Where H[i] is constructed from the elementary reflectors as computed by Dsytrd.
+// Please see the documentation for Dsytrd for more information.
+//
+// tau must have length at least n-1, and Dorgtr will panic otherwise.
+//
+// Work is temporary storage, and lwork specifies the usable memory length. At minimum,
```
Comment (kortschak) [inline native/dorgtr.go]: I think we need to come up with a decision about case for tau, work etc.

File: native/dorgtr.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+	if n == 0 {
+		return
+	}
+	if upper {
+		// Q was determined by a call to Dsytrd with uplo == blas.Upper.
+		// Shift the vectors which define the elementary reflectors one column
+		// to the left, and set the last row and column of Q to those of the unit
+		// matrix.
+		for j := 0; j < n-1; j++ {
+			for i := 0; i <= j-1; i++ {
```
Comment (kortschak) [inline native/dorgtr.go]: `i < j`?

File: native/dorgtr.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+		a[(n-1)*lda+n-1] = 1
+
+		// Generate Q[0:n-1, 0:n-1].
+		impl.Dorgql(n-1, n-1, n-1, a, lda, tau, work, lwork)
+	} else {
+		// Q was determined by a call to Dsytrd with uplo == blas.Upper.
+		// Shift the vectors which define the elementary reflectors one column
+		// to the right, and set the first row and column of Q to those of the unit
+		// matrix.
+		for j := n - 1; j >= 1; j-- {
```
Comment (kortschak) [inline native/dorgtr.go]: `j > 0`?

File: native/dsytrd.go [hunk 10]
```diff
@@ -135,7 +135,7 @@ func (impl Implementation) Dsytrd(uplo blas.Uplo, n int, a []float64, lda int, d
 			// Update the unreduced submatrix A[i+ib:n, i+ib:n], using an update
 			// of the form A = A + V*W^T - W*V^T.
 			bi.Dsyr2k(uplo, blas.NoTrans, n-i-nb, nb, -1, a[(i+nb)*lda+i:], lda,
-				work[nb+i:], ldwork, 1, a[(i+nb)*lda+i+nb:], lda)
+				work[nb*ldwork:], ldwork, 1, a[(i+nb)*lda+i+nb:], lda)
```
Comment (kortschak) [inline native/dsytrd.go]: Comment on this fix in the commit message.

**Comment by kortschak** [comment]
LGTM after comments addressed.

**Comment by btracey** [comment]
In this particular case it's not a prng thing, at least not in the sense of the other tests that fail when the random numbers are different. In this case the code was wrong and discovered in the first "real" input (as determined by dsytrd). There's something about the real input which triggers the bug that my direct tests do not

File: native/dorgtr.go (hunk 1 above)
Comment (btracey) [inline native/dorgtr.go]: Done.

File: native/dorgtr.go (hunk 2 above)
Comment (btracey) [inline native/dorgtr.go]: Done.

File: native/dorgtr.go (hunk 3 above)
Comment (btracey) [inline native/dorgtr.go]: Sorry. Done.

File: native/dorgtr.go (hunk 4 above)
Comment (btracey) [inline native/dorgtr.go]: yep.

File: native/dorgtr.go (hunk 5 above)
Comment (btracey) [inline native/dorgtr.go]: Hard to find bug. Removed.

File: native/dorgtr.go (hunk 6 above)
Comment (btracey) [inline native/dorgtr.go]: I don't think that's the right order. It seems much more likely that someone would accidentally input a bad value for lwork (0 or something). Given the len() command in go there is no reason for the input caller to have a lesser value, so if that condition is true something weird is happening and it should be alerted first.

File: native/dorgtr.go (hunk 7 above)
Comment (btracey) [inline native/dorgtr.go]: Made lowercase now like in the recent batch of functions.

File: native/dorgtr.go (hunk 8 above)
Comment (btracey) [inline native/dorgtr.go]: Done.

File: native/dorgtr.go (hunk 9 above)
Comment (btracey) [inline native/dorgtr.go]: Done.

**Comment by kortschak** [comment]
The dsytrd change commit comment was missed.

**Comment by btracey** [comment]
I don't understand how. I edited the commit message when I rebased and also
in the github api

**Comment by kortschak** [comment]
It's in the merge commit, but it should be in the actual commit. During the rebase this would have happened (or not).

**Comment by kortschak** [comment]
Do you know how the fortran equivalents behave with the failing case?

**Comment by btracey** [comment]
Not specifically, but I'd imagine they would similarly fail. I think it may be the same issue that I had when trying to get Dgesvd to work. That is, Go is fine with slicing beyond the end of the matrix, and so if there is an indexing error plus data originally being part of the same array there is no panic. This relates to your valid issue that we really should be using full slicing on the slices and not just pointer movement.

**Pull request merged**